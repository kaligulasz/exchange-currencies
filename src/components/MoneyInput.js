import React, { useState, useEffect } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

const Input = styled.input`
  border: 0;
  background: none;
  height: 3rem;
  text-align: right;
  font-size: 2rem;
  color: ${props => props.theme.color.white};
  position: relative;
  outline: none;
  max-width: 10rem;
  
  ::placeholder {
    color: rgba(255, 255, 255, 0.3);
  }
`;

const MinusIcon = styled.div`
  width: 0.7rem;
  height: 3px;
  margin-right: -0.5rem;
  background: ${props => props.theme.color.white};
  border-radius: 0.07rem;
`;

const Wrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: flex-end;
`;


const MoneyInput = ({ onChange, maxValue }) => {
  const [inputValue, setInputValue] = useState('');

  useEffect(() => {
    if (maxValue < inputValue) {
      setInputValue('');
      onChange('');
    }
  }, [maxValue]);

  const validateDotDuplication = value => [...value].filter(char => char.includes('.')).length <= 1;

  const validateTwoDigitsAfterDot = (value) => {
    const { 1: afterDot = '' } = value.split('.');

    return afterDot.length < 3;
  };

  const checkValidation = (event) => {
    const { value } = event.target;
    const regex = /^[0-9.]+$/;

    if (
      (value === '' || regex.test(value))
      && validateDotDuplication(value)
      && validateTwoDigitsAfterDot(value)
      && maxValue >= value
    ) {
      onChange(value);
      setInputValue(value);
    }
  };

  return (
    <Wrapper>
      {inputValue && <MinusIcon />}
      <Input
        onChange={checkValidation}
        size={inputValue.length}
        type="text"
        value={inputValue}
        placeholder="amount"
        autofocus="true"
      />
    </Wrapper>
  );
};

MoneyInput.propTypes = {
  onChange: PropTypes.func.isRequired,
  maxValue: PropTypes.number.isRequired,
};

export default MoneyInput;
