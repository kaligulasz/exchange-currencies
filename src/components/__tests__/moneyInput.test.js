import React from 'react';
import { shallow, mount } from 'enzyme';
import { ThemeProvider } from 'styled-components';
import 'jest-styled-components';

import theme from '../../common/themeVariables';

import MoneyInput from '../MoneyInput';

let wrapper;

beforeEach(() => {
  wrapper = mount(
    <ThemeProvider theme={theme}>
      <MoneyInput maxValue={20} onChange={() => {}} />
    </ThemeProvider>,
  );
});

describe('<MoneyInput />', () => {
  it('renders <MoneyInput /> with passed props', () => {
    const component = shallow(
      <MoneyInput maxValue={20} onChange={() => {}} />,
    );

    expect(component).toMatchSnapshot();
  });

  it('check input validation', () => {
    const event = { target: { name: 'money-input', value: '20' } };
    wrapper.find('input').simulate('change', event);
    const input = wrapper.find('input');

    expect(input.props().value).toEqual('20');
  });

  it('check input validation: two numbers after dot', () => {
    const event = { target: { name: 'money-input', value: '20.000' } };
    wrapper.find('input').simulate('change', event);
    const input = wrapper.find('input');

    expect(input.props().value).toEqual('');
  });

  it('check input validation: max value', () => {
    const event = { target: { name: 'money-input', value: '31' } };
    wrapper.find('input').simulate('change', event);
    const input = wrapper.find('input');

    expect(input.props().value).toEqual('');
  });

  it('check input validation: max two zero characters before dot', () => {
    const event = { target: { name: 'money-input', value: '00000.' } };
    wrapper.find('input').simulate('change', event);
    const input = wrapper.find('input');

    expect(input.props().value).toEqual('');
  });

  it('check input validation: only numbers are allowed', () => {
    const event = { target: { name: 'money-input', value: 'someText' } };
    wrapper.find('input').simulate('change', event);
    const input = wrapper.find('input');

    expect(input.props().value).toEqual('');
  });

  it('check input validation: only one dot is allowed', () => {
    const event = { target: { name: 'money-input', value: '10.10.0' } };
    wrapper.find('input').simulate('change', event);
    const input = wrapper.find('input');

    expect(input.props().value).toEqual('');
  });
});
