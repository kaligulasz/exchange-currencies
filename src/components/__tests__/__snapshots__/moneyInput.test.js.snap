// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MoneyInput /> renders <MoneyInput /> with passed props 1`] = `
<MoneyInput__Wrapper>
  <MoneyInput__Input
    autoFocus={true}
    name="money-input"
    onChange={[Function]}
    placeholder="amount"
    size={0}
    type="text"
    value=""
  />
</MoneyInput__Wrapper>
`;
